### PR TITLE
Add IndirectCall service to JITBuilder

### PR DESCRIPTION
**Pull request by cathyzhyi**
Signed-off-by: Yi Zhang <yizhang@ca.ibm.com>

### CODE REVIEW

File: compiler/il/OMRILOps.hpp [hunk 1]
```diff
@@ -1180,6 +1180,23 @@ class ILOpCode
       return TR::BadILOp;
       }
 
+   static TR::ILOpCodes getIndirectDirectCall(TR::DataType type)
```
Comment (mstoodle) [inline compiler/il/OMRILOps.hpp]: Should this be getIndirectCall() maybe?

File: compiler/ilgen/IlBuilder.cpp [hunk 2]
```diff
@@ -1325,24 +1325,38 @@ IlBuilder::GreaterThan(TR::IlValue *left, TR::IlValue *right)
    return returnValue;
    }
 
-TR::IlValue *
-IlBuilder::Call(const char *functionName, int32_t numArgs, ...)
+static TR::IlValue **processCallArgs(TR::Compilation *comp, int numArgs, va_list args)
```
Comment (mstoodle) [inline compiler/ilgen/IlBuilder.cpp]: Why is this a static function?

File: compiler/ilgen/IlBuilder.cpp [hunk 3]
```diff
@@ -1355,7 +1369,7 @@ IlBuilder::Call(const char *functionName, int32_t numArgs, TR::IlValue ** argVal
 
    // treat as "Static" (so no receiver expected) and use a direct call opcode
```
Comment (mstoodle) [inline compiler/ilgen/IlBuilder.cpp]: This comment should be updated

File: compiler/ilgen/IlBuilder.cpp [hunk 4]
```diff
@@ -1365,16 +1379,16 @@ IlBuilder::Call(const char *functionName, int32_t numArgs, TR::IlValue ** argVal
       callNode->setAndIncChild(childIndex++, loadValue(arg));
       }
 
+   // call has side effect and needs to be anchored under a treetop
+   genTreeTop(callNode);
+
```
Comment (mstoodle) [inline compiler/ilgen/IlBuilder.cpp]: Good catch!

File: compiler/ilgen/IlBuilder.cpp [hunk 5]
```diff
@@ -1336,56 +1336,90 @@ IlBuilder::GreaterThan(TR::IlValue *left, TR::IlValue *right)
...
    if (returnType != TR::NoType)
       {
       TR::IlValue *returnValue = newValue(callNode->getDataType());
-      storeNode(returnValue, callNode);
+      genTreeTop(TR::Node::createStore(returnValue, callNode));
```
Comment (mstoodle) [inline compiler/ilgen/IlBuilder.cpp]: Was this change necessary? I suppose it creates one less block...

File: compiler/ilgen/IlBuilder.cpp [hunk 6]
```diff
@@ -1336,56 +1336,90 @@ IlBuilder::GreaterThan(TR::IlValue *left, TR::IlValue *right)
...
+ * \param computedVirtualCallAddress
+ *    It's normally an address loaded from the virtual function table
+ * \param numArgs 
+ *    Number of real arguments from the method signature
+ * \param ... 
+ *    The real arguments from the method signature
+ */
 TR::IlValue *
-IlBuilder::Call(const char *functionName, int32_t numArgs, TR::IlValue ** argValues)
+IlBuilder::ComputedVirtualCall(const char *functionName, TR::IlValue *computedVirtualCallAddress, int32_t numArgs, ...)
```
Comment (mstoodle) [inline compiler/ilgen/IlBuilder.cpp]: I think an easier notation would be:
   ComputedCall(const char *functionName, int32_t numArgs, ...)

then pull off the first arg as computedCallAddress. That makes the calling sequence a bit more natural too:
b->ComputedCall("foo", 2,
b->   Load("the target entry point"),
b->   Load("parameter1"));

as opposed to the mix:
b->ComputedCall("foo",
b->   Load("the target entry point"), 1,
b->   Load("parameter 1"));

I think I prefer the first one without the numArgs value (1) "lost" in the middle.

I'm also not sure there's anything strictly "virtual" about this call, since the "virtualness" is all done by the client of the interface. That's why I removed "Virtual" from the name in the above examples.

File: compiler/x/amd64/codegen/AMD64SystemLinkage.cpp [hunk 7]
```diff
@@ -724,6 +724,7 @@ TR::Register *
 TR_AMD64SystemLinkage::buildIndirectDispatch(TR::Node *callNode)
    {
    TR::SymbolReference *methodSymRef = callNode->getSymbolReference();
+   TR_ASSERT(methodSymRef->getSymbol()->castToMethodSymbol()->isComputedStatic(), "system linkage only supports computed static call for now %p\n", callNode);
```
Comment (mstoodle) [inline compiler/x/amd64/codegen/AMD64SystemLinkage.cpp]: Should we provide this assert in the system linkage implementations for other platforms too?

Comment (cathyzhyi) [inline compiler/x/amd64/codegen/AMD64SystemLinkage.cpp]: Yea, I agree. Added the same assert for ppc, x86-32 bits doesn't support indirect call at all and there is an assertion for that already. I also changed the assert to isComputed() because for codegen computedStatic and computedVirtual are the same. It is regular virtual call vs computed call that matters.

Comment (cathyzhyi) [inline compiler/x/amd64/codegen/AMD64SystemLinkage.cpp]: Just confirmed with Filip and added the Z part as well.

**Comment by andrewcraik** [comment]
@cathyzhyi @mstoodle is this ready to merge now? if so a rebase and merge to avoid the item rotting is probably a good idea.

**Comment by cathyzhyi** [comment]
done rebased to the latest version.

**Comment by mstoodle** [comment]
Sorry for the delay. Thanks @cathyzhyi and @andrewcraik !
